File: core/weapons/plasmaCanon.js
import { randomNum } from "../utils/random.js";
import Projectile from "./projectile.js";
import Minature from "./minature.js";
import WeaponManager from "./manager.js";
import { assets } from "../assets/main.js";

export default class PlasmaCanon extends Projectile {
    constructor({ x, y, angle, ship, color, name = "Plasma Canon", img = assets?.images?.plasma, acceleration = 25000, width = 10, height = 45, damage = 10, range = 5000, speed = 10, fireRate = 0.002, energyCost = 3000, }) {
        super({
            name,
            speed: speed,
            acceleration,
            x: x,
            y: y,
            width,
            height,
            angle,
            damage,
            range,
            fireRate,
            energyCost,
            ship,
            color,
            img,
        });
    }

    explode(radius = 1000) {
        if (!this.active) return;
        for (let i = 0; i < 100; i++) {
            const prop = {
                x: this.x - Math.sin(this.angle) * this.width,
                y: this.y - Math.cos(this.angle) * this.height,
                angle: randomNum(-Math.PI, Math.PI),
                speed: randomNum(radius / 2, radius),
                ship: this.ship,
                range: randomNum(radius * 0.1, radius / 2),
                color: "yellow"
            }


            WeaponManager.fire(Minature, prop, true)
        }
    }

    travelEnd() {
        this.explode()
    }

    colide() {
        this.explode();
    }

    // update(dt, manager) {
    //     super.update(dt, manager)
    // }
}

File: void_front/game/core/weapons/plasma_canon.py
from __future__ import annotations

import numpy as np
from game.core.weapons.manager import WeaponManager
from game.core.weapons.minature import Minature
from game.core.weapons.projectile import Projectile


class PlasmaCanon(Projectile):
    def __init__(self, options, force=False):
        defaults = {
            "name": "Plasma Canon",
            "speed": options.get("speed", 10),
            "acceleration": options.get("acceleration", 25000),
            "width": options.get("width", 10),
            "height": options.get("height", 45),
            "damage": options.get("damage", 10),
            "range": options.get("range", 5000),
            "fireRate": options.get("fireRate", 0.002),
            "energyCost": options.get("energyCost", 3000),
            "x": options.get("x"),
            "y": options.get("y"),
            "angle": options.get("angle"),
            "ship": options.get("ship"),
            "color": options.get("color"),
            "img": options.get("img"),
        }
        super().__init__(defaults, force)

    def explode(self, radius=1000):
        if not self.active:
            return
        for _ in range(100):
            prop = {
                "x": self.x - np.sin(self.angle) * self.width,
                "y": self.y - np.cos(self.angle) * self.height,
                "angle": np.random.uniform(-np.pi, np.pi),
                "speed": np.random.uniform(radius / 2, radius),
                "ship": self.ship,
                "range": np.random.uniform(radius * 0.1, radius / 2),
                "color": "yellow",
            }
            WeaponManager.fire(Minature, prop, True)

    def travelEnd(self):
        self.explode()

    def colide(self):
        self.explode()
